src/net_processing.cpp

```diff
@@ -3807,15 +3837,14 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             vRecv >> nonce;
 
             // Only process pong message if there is an outstanding ping (old ping without nonce should never pong)
-            if (pfrom.nPingNonceSent != 0) {
-                if (nonce == pfrom.nPingNonceSent) {
+            if (peer->m_ping_nonce_sent != 0) {
+                if (nonce == peer->m_ping_nonce_sent) {
                     // Matching pong received, this ping is no longer outstanding
                     bPingFinished = true;
-                    const auto ping_time = ping_end - pfrom.m_ping_start.load();
+                    const auto ping_time = ping_end - peer->m_ping_start.load();
                     if (ping_time.count() >= 0) {
```

Comment:
I think `ping_time.count == 0` is expected if you do `setmocktime X`, then a ping goes out, and the pong is received before you call `setmocktime Y`.

Reply:
Leaving this so as not to change behaviour

Reply:
(This was meant to be a reply to https://github.com/bitcoin/bitcoin/pull/20721#discussion_r572489932 except github wasn't let me comment there)